src/main/java/org/jabref/gui/maintable/ExtractReferencesAction.java

```diff
@@ -68,34 +102,113 @@ public void execute() {
                                                               .withInitialMessage(Localization.lang("Processing PDF(s)"));
 
             task.onFailure(dialogService::showErrorDialogAndWait);
 
             ImportEntriesDialog dialog = new ImportEntriesDialog(stateManager.getActiveDatabase().get(), task);
-            dialog.setTitle(Localization.lang("Extract References"));
+            String title;
+            if (online) {
+                title = Localization.lang("Extract References (online)");
+            } else {
+                title = Localization.lang("Extract References (offline)");
+            }
+            dialog.setTitle(title);
             dialogService.showCustomDialogAndWait(dialog);
         });
     }
+
+    private @NonNull Callable<ParserResult> getParserResultCallableOffline(BibDatabaseContext databaseContext, List<BibEntry> selectedEntries) {
+        return () -> {
+            BibEntry currentEntry = selectedEntries.getFirst();
+            List<Path> fileList = FileUtil.getListOfLinkedFiles(selectedEntries, databaseContext.getFileDirectories(preferencesService.getFilePreferences()));
+
+            // We need to have ParserResult handled at the importer, because it imports the meta data (library type, encoding, ...)
+            ParserResult result = bibliographyFromPdfImporter.importDatabase(fileList.getFirst());
+
+            // subsequent files are just appended to result
+            Iterator<Path> fileListIterator = fileList.iterator();
+            fileListIterator.next(); // skip first file
+            extractReferences(fileListIterator, result, currentEntry);
+
+            // handle subsequent entries
+            Iterator<BibEntry> selectedEntriesIterator = selectedEntries.iterator();
+            selectedEntriesIterator.next(); // skip first entry
+            while (selectedEntriesIterator.hasNext()) {
+                currentEntry = selectedEntriesIterator.next();
+                fileList = FileUtil.getListOfLinkedFiles(List.of(currentEntry), databaseContext.getFileDirectories(preferencesService.getFilePreferences()));
+                fileListIterator = fileList.iterator();
+                extractReferences(fileListIterator, result, currentEntry);
+            }
+
+            return result;
+        };
+    }
+
+    private void extractReferences(Iterator<Path> fileListIterator, ParserResult result, BibEntry currentEntry) {
+        while (fileListIterator.hasNext()) {
+            result.getDatabase().insertEntries(bibliographyFromPdfImporter.importDatabase(fileListIterator.next()).getDatabase().getEntries());
+        }
+
+        StringJoiner cites = new StringJoiner(",");
+        int count = 0;
+        for (BibEntry importedEntry : result.getDatabase().getEntries()) {
+            count++;
+            Optional<String> citationKey = importedEntry.getCitationKey();
+            if (citationKey.isPresent()) {
```

Comment:
citationKey.map(cites:add).orElseGet( () ->

Reply:
Not sure if new code is more readable --> "orElseGet" result needs to be added to the list, too. Uses outer variable "count", which is non final. I needed to wrap in anonymous object.

Reply:
then better use the original code